### PR TITLE
Use multiple Tigris instances to run tests

### LINKED ISSUE #1568: Rework integration test setup and auth
### What should be done?

At the moment, we fetch flag values but each test can't assign value for the flag. Additionally `-handler` flag is used incorrectly for validator.

* Validators should be used only for target. When we run compatibility tests with target=tigris and compat=MongoDB, we get validators for `tigris` handler for both databases and then ignore MongoDB's error. In general, `-handler`'s flag value is used incorrectly. https://github.com/FerretDB/FerretDB/blob/f0ca7811ed2aae21ce5063bd3cab533d56cf6cfb/integration/setup/setup_compat.go#L131
* Integration setup uses flag's value but unable to specify flags used per test cases. Enable test cases to test authentication. https://github.com/FerretDB/FerretDB/blob/f0ca7811ed2aae21ce5063bd3cab533d56cf6cfb/integration/setup/common.go#L231
* Test different flags values of env data `TestEnvData`.
* Test different flag values of `postgresql-url` in `TestCommandsAuthenticationSASLStart` 
* Provider is compatible with both compat and target, but validator is only compatible with target. Fix this definition https://github.com/FerretDB/FerretDB/blob/62039bf08d8cb4cd19bbc39c0c75008f13e3a48c/integration/shareddata/shareddata.go#L33-L39
* pg uri used for testing is set from creating a new listener or using existing connection on port. Rework on how uri is built and used. Currently attempting all possible combinations. Probably this wouldn't help for speed. https://github.com/FerretDB/FerretDB/blob/cc7b47db1efce9c164f5922eb564ea6b7a8592ef/integration/setup/common.go#L170-L181
* Remove code like https://github.com/FerretDB/FerretDB/blob/56a2dd17b77bcb6afe1fc64b592a7985c4df15b7/integration/create_test.go#L70-L81

Keep note that maybe we may want test data per provider.
### Where?

See references to this issue in code.

DoD:
- [x] in compat test, validators are correctly applied for target but not for compat.
- [x] tests added for `TestCommandsAuthenticationSASLStart`
  - [x] authentication with PLAIN
  - [ ] authentication with invalid credential
  - [x] authentication with valid credential
- [x] uri for specific target/compat is being built.
- [x] validator is not used for compat
- [ ] address issue numbers mentioned in the code

---

- [x] rework validations in shareddata
- [x] fix `task env-data` (#2200)
- [x] fix test script

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #1568.

Use 4 additional instances for running tests for tigris exposed on ports 8082, 8083, 8085 and 8086. Skipping 8084 because it's used by something else.

On CI, this change appears to runs tigris tests in 6 to 7 minutes. I checked recent merges to main which tigris tests appear to run in 10 to 11min. I don't know how much more $$ it might use.
 
## Readiness checklist

* [x] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [ ] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/setup/startup.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// startupInitializer creates an instance of startupInitializer.
+func newStartupInitializer() *startupInitializer {
+	nPortCalls := uint64(0)
+	return &startupInitializer{nPortCalls: &nPortCalls}
+}
+
+// getNextTigrisPort gets the next port number of Tigris to be used
+// for testing in Round Robin fashion.
```
Comment (noisersup) [inline integration/setup/startup.go:38]: Nice usage!

**Review by noisersup — APPROVED** [review]
Tigris tests faster than PostgreSQL? :laughing: I guess it's because of number of skips

LGTM!

File: integration/setup/startup.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// Package setup provides integration tests setup helpers.
+package setup
+
+import (
+	"sync/atomic"
+)
+
+// ports are available port of Tigris.
+var ports = []uint16{8081, 8082, 8083, 8085, 8086}
```
Comment (rumyantseva) [inline integration/setup/startup.go:23]: I like this idea, but I'm a bit skeptical about these ports being defined here as the exact number of exact ports and not being configurable.
Is there a plan to make those things configurable (like `tigris-url` itself) - maybe in separate PRs?

**Review by rumyantseva — APPROVED** [review]
I like the idea! Asked a question about configuration.

File: integration/setup/startup.go (hunk 2 above)
Comment (chilagrow) [reply integration/setup/startup.go:23]: I've looked at it briefly and decided I would not do it for this, currently there were several places where ports are defined, env tool, docker-compose.yml, and within integrations tests. So it's nice to have configurable but also it touches more than the integration test parameter.  For this PR, I'm gonna do quick win. :see_no_evil:

Comment (rumyantseva) [reply integration/setup/startup.go:23]: Oh, `envtool` is a nice catch as we wait for `8081` only, and not for the other ports...  
I'm ok to merge it as is to have a quick win, and maybe we can reconsider this and make it configurable after #1857 (so, it'll be easier to deal with configuration).

**Review by w84thesun — APPROVED** [review]
👍

File: cmd/envtool/envtool.go [hunk 3]
```diff
@@ -151,33 +151,36 @@ func setupPostgresSecured(ctx context.Context, logger *zap.SugaredLogger) error
 func setupTigris(ctx context.Context, logger *zap.SugaredLogger) error {
 	logger = logger.Named("tigris")
 
-	err := waitForPort(ctx, logger, 8081)
-	if err != nil {
-		return err
-	}
+	// Skip 8084 because since it is used by userland proxy.
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:154]: By what "userland proxy"?

File: integration/setup/startup.go (hunk 2 above)
Comment (AlekSi) [reply integration/setup/startup.go:23]: I think we should start paying a bit more attention to our code comments and not leave explanations on GitHub only, where they are unlikely to be found.

In this case, cross-linking those ports to the Docker Compose configuration would significantly improve [understanding](https://www.osnews.com/story/19266/wtfsm/).

Comment (chilagrow) [reply integration/setup/startup.go:28]: Yeah agree. I've added comments. I also foresee the number of instances are just a guess and need adjusting hence a new issue https://github.com/FerretDB/FerretDB/issues/1887

**Review by w84thesun — DISMISSED** [review]
👍

**Review by noisersup — APPROVED** [review]
:shipit:

**Review by rumyantseva — APPROVED** [review]
LGTM

File: integration/setup/startup.go (hunk 2 above)
Comment (AlekSi) [reply integration/setup/startup.go]: See bd05d83c9d5df8e36f41de8aff7a95bb60ed215b

**Review by AlekSi — APPROVED** [review]

**Comment by chilagrow** [comment]
Thank you for the clean up @AlekSi :hugs:

**Pull request merged**